dropshot/src/error.rs

```diff
@@ -50,6 +50,89 @@ use serde::Serialize;
 use std::error::Error;
 use std::fmt;
 
+pub use crate::extractor::ExtractorError;
+pub use crate::extractor::MultipartBodyError;
+pub use crate::extractor::QueryError;
+pub use crate::extractor::StreamingBodyError;
+pub use crate::extractor::TypedBodyError;
+pub use crate::http_util::PathError;
+pub use crate::router::RouterError;
+
+#[derive(Debug, thiserror::Error)]
+pub enum ServerError {
```

Comment:
This is the top-level representation of "all errors generated from within dropshot" which the user can construct their own error responses from using the server-level error handler. Alternatively, if the error handler is not overridden, we will just convert into the stock `HttpError` for you by default.

I also considered calling this `DropshotError` --- open to being convinced either way.

Reply:
Something else I considered was having a `fn suggested_status_code()` or something on this and the nested child errors, which would return the same status code we use when converting them to `HttpError`s. This would be in order to help users construct their own error response types without having to match every possible variant of these errors. You could then use a combination of the `suggested_status_code()` and `to_string()` methods to construct a user error type, and potentially match specific variants and handle them differently.

But, this API can be added in a separate PR.